Add Support for transposed convolution with Padding Mode 'same' and 'valid'

Alvaro-Kothe opened the pull request:
This pull request adds support for `'same'` and `'valid'` padding modes for transposed convolutions.

### Implementation Details for `'same'` padding:

1. Compute the minimum required padding for the left and right sides.
2. Perform the convolution using the computed padding.
3. Adjust the output so that its size matches the input size.

### Notes:

* The padding calculation is based on the [JAX implementation](https://github.com/jax-ml/jax/blob/6a6a13dcfe5f6dbca4155410dd32dfc2d5bf1cec/jax/_src/lax/convolution.py#L253).
* For asymmetric convolutions, I slice one extra value on the left side.
* `'valid'` padding is equivalent to `padding=0`.
* `'same'` padding is **not supported** when:

  * The stride is not equal to 1 (`stride != 1`), as it is also unsupported in traditional convolutions, and I could not find any bibliographic references clarifying how it could be implemented in this context.
  * `output_padding != 0`, since this seems to conflict with the intent of `'same'` padding.

Close #80301; Close #3867


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @jerryzh168

Alvaro-Kothe committed:
feat(ConvTranspose): add support for padding as `kSame` and `kValid`

Adds support for `padding=torch::k{Same,Valid}` for transposed
convolutions in C++ API.

Alvaro-Kothe committed:
refactor: run lintrunner

shoumikhin commented:
@pytorchbot rebase

Alvaro-Kothe committed:
fix: move `min_pad` to vector

Alvaro-Kothe commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -1109,6 +1111,211 @@ at::Tensor conv3d_padding_symint(
+        conv_transpose_same_mode_padding_lr(effective_kernel, dim_stride);
+
+    auto min_pad = pad.second < pad.first ? pad.second : pad.first;
+    common_pad_lr.push_back(min_pad);
```
Done

Alvaro-Kothe commented:
Could someone please trigger the CI?
Soft ping to: @Skylion007 @albanD @shoumikhin. Thank you in advance!

Alvaro-Kothe commented:
@pytorchbot rebase

Alvaro-Kothe commented:
@pytorchbot rebase

albanD reviewed (commented):
@jbschlosser Can you take a quick look at the updated native_function.yaml changes.
I don't recall off the top of my head what we do for regular convolution there. But passing strings around sound unexpected for sure.

Alvaro-Kothe commented:
@albanD Thanks for taking a look. Here is regular convolution using padding as a string.

https://github.com/pytorch/pytorch/blob/8f32adc90a7fee83583c9ba89dbdfabb317e0452/aten/src/ATen/native/native_functions.yaml#L1753-L1766

jbschlosser commented:
> @jbschlosser Can you take a quick look at the updated native_function.yaml changes.
I don't recall off the top of my head what we do for regular convolution there. But passing strings around sound unexpected for sure.

Arguably, we shouldn't have passed strings in the original conv*d padding overload signatures (added in #45667). But as @Alvaro-Kothe pointed out, this is indeed the case. These highest-level overloads are composite and strings don't exist in any conv-related op signatures at lower levels.

albanD reviewed (commented):
The code sounds ok. I would expect to see a python side for this as well as python tests?

Alvaro-Kothe committed:
fix: remove private function from publicAPI.json

Alvaro-Kothe commented on test/allowlist_for_publicAPI.json:
```diff
@@ -1129,6 +1129,7 @@
     "_convert_indices_from_csr_to_coo",
     "_convolution",
     "_convolution_mode",
+    "_convolution_transpose_mode",
```
I removed it. 

I think this rule applies to `_convolution_mode` too, but removing it goes out of the scope of this PR.

albanD replied:
Yes we can remove these others in another PR

Alvaro-Kothe commented on torch/csrc/api/include/torch/nn/functional/conv.h:
```diff
@@ -166,12 +166,23 @@ inline Tensor conv_transpose1d(
     IntArrayRef dilation) {
-  return torch::conv_transpose1d(
-      input, weight, bias, stride, padding, output_padding, groups, dilation);
+  return std::visit(
```
Additionally, `conv*d` does something similar.

https://github.com/pytorch/pytorch/blob/7f3dc45300b9c5a3f3709f1cc810f46fbdd14ec4/torch/csrc/api/include/torch/nn/functional/conv.h#L24-L38

albanD replied:
Ho I see, I'm not that familiar with variant. That sounds good.

Alvaro-Kothe committed:
feat(conv_transpose): add support for padding={same,valid} in Python

albanD reviewed (commented):
Thanks for the update. It looks great.
Just a quick question to Joel about testing.

Alvaro-Kothe committed:
test(conv_tranpose): add inputs to test padding as string

Alvaro-Kothe commented on test/nn/test_convolution.py:
```diff
@@ -2178,6 +2178,549 @@ def test_conv3d_valid_padding_backward(self, device, dtype):
             check_fwd_over_rev=check_forward_ad,
         )
 
+    # conv_transpose*d_padding_* tests adapted from conv*d_padding_*
```
I added new entries for `sample_inputs_conv_transpose{1,2,3}d` on 87d0c58 and verified the tests by running `test/test_meta.py`.

Alvaro-Kothe commented:
I reproduced the [dynamo](https://github.com/pytorch/pytorch/actions/runs/18281843725/job/52076095467#step:25:6410) bug locally, and I will try to figure out what is wrong in the `complex128` part to fix it.

albanD commented:
Sounds good!
Don't hesitate to ask here if there is anything that looks too weird haha
In this case, you might simply need to add "decomposition"

albanD commented:
Ho the fact that a rename fixes it means there is some schenanigans to skip/ignore it, most likely one of the weird way dynamo skip happens, but I can't find it now.
You can add `@skipIfTorchDynamo("Doesn't work similar to conv3d")` to the test to skip it in this case.

Alvaro-Kothe committed:
test(conv_transpose3d): skip on dynamo

Test fails, similar to `conv3d`, on dynamo with error

```
TypeError: add(): argument 'input' (position 1) must be Tensor, not NoneType
```

albanD reviewed (approved):
Thanks!
Approving and we can merge once CI is all green!

Alvaro-Kothe committed:
test: add dynamo skip for transpose 3d valid

Alvaro-Kothe committed:
test(conv_transposeNd): make conv_tranpose_ref handle padding as a string

Alvaro-Kothe commented:
@pytorchbot rebase

Alvaro-Kothe commented:
@pytorchbot rebase
